codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State implements PriorityQueueNode<State> {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        /**
+         * Count of nodes rooted at this sub tree with {@link #isActive} equal to {@code true}.
+         */
+        int isActiveCountForTree;
+        private int priorityQueueIndex = INDEX_NOT_IN_QUEUE;
+        /**
+         * An estimate of when this node should be given the opportunity to write data.
+         */
+        long pseudoTimeToWrite;
+        /**
+         * A pseudo time maintained for immediate children to base their {@link pseudoTimeToSend} off of.
+         */
+        long pseudoTime;
+        long totalQueuedWeights;
+        boolean isActive;
+
+        State(Http2Stream stream) {
+            this(stream, 0);
+        }
+
+        State(Http2Stream stream, int initialSize) {
+            this.stream = stream;
+            queue = new PriorityQueue<State>(initialSize);
+        }
+
+        void write(int numBytes, Writer writer) throws Http2Exception {
+            try {
+                writer.write(stream, numBytes);
+            } catch (Throwable t) {
+                throw connectionError(INTERNAL_ERROR, t, "byte distribution write error");
+            }
+        }
+
+        void isActiveCountChangeForTree(int increment) {
+            assert isActiveCountForTree + increment >= 0;
+            isActiveCountForTree += increment;
+            if (!stream.isRoot()) {
+                State pState = state(stream.parent());
+                if (isActiveCountForTree == 0) {
+                    pState.remove(this);
+                } else if (isActiveCountForTree - increment == 0) { // if frame count was 0 but is now not, then queue.
+                    pState.offerAndInitializePseudoTime(this);
+                }
+                pState.isActiveCountChangeForTree(increment);
+            }
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean isActive) {
+            if (this.isActive != isActive) {
+                isActiveCountChangeForTree(isActive ? 1 : -1);
+                this.isActive = isActive;
+            }
+
+            streamableBytes = newStreamableBytes;
+        }
+
+        /**
+         * Assumes the parents {@link #totalQueuedWeights} includes this node's weight.
+         */
+        void updatePseudoTime(State parentState, int nsent, long totalQueuedWeights) {
+            assert stream.id() != CONNECTION_STREAM_ID && nsent >= 0;
+            // If the current pseudoTimeToSend is greater than parentState.pseudoTime then we previously over accounted
+            // and should use parentState.pseudoTime.
+            pseudoTimeToWrite = (long) (min(pseudoTimeToWrite, parentState.pseudoTime) +
+                                    nsent * ((double) totalQueuedWeights / stream.weight()));
```

Comment:
Cast `nsent` to long and then get rid of the `double` cast and parenthesis?

We shouldn't be concerned with overflow, since that would mean that if each stream sends `nsent`, `pseudoTime` itself would overflow, _even if `pseudoTime == 0` right now_.

Using `double` also loses precision unless we cast the product to long _before_ the addition (because in general `pseudoTime` can be "large" since it accumulates over the lifetime of the connection, whereas the other values would be relatively small).